app/api/src/players/players.service.ts

```diff
@@ -42,4 +43,129 @@ export class PlayersService {
 		}
 		return player;
 	}
+
+	async findTop(topPlayersQuery: TopPlayersQueryDto) {
+		const { stat: hitsStat, gameType, minGames, minShots, limit } = topPlayersQuery;
+
+		const shotsStat = {
+			discDmgHitsTG: 'discShotsFiredTG',
+			discHitsTG: 'discShotsFiredTG',
+			discMATG: 'discShotsFiredTG',
+			laserHitsTG: 'laserShotsFiredTG',
+			laserMATG: 'laserShotsFiredTG',
+			cgHitsTG: 'cgShotsFiredTG',
+			shockHitsTG: 'shockShotsFiredTG'
+		}[hitsStat]
+
+		// Possibly make this a query param at some point?
+		const excludeDiscJumps = shotsStat === 'discShotsFiredTG';
+
+		const hitsValue = '(game.stats->:hitsStat->>0)::integer';
+		const shotsValue = '(game.stats->:shotsStat->>0)::integer';
+		const discJumpsValue = "(game.stats->'discJumpTG'->>0)::integer";
+		const killerDiscJumpsValue = "(game.stats->'killerDiscJumpTG'->>0)::integer";
+
+		// pgSQL doesn't let you reference aliased selections you've made in the
+		// same select statement, so unfortunately these computed JSON values are
+		// repeated in a couple places. Using template strings here to easily repeat
+		// the same expression instead of having multiple nested subqueries. Rest
+		// assured, no user-supplied values are ever interpolated, those are all
+		// parameterized instead.
```

Comment:
Basically what I'd like to do is:

```
SELECT
  (some expression) AS hits,
  (some expression) AS shots,
  (hits / shots) AS accuracy
```

but pgSQL doesn't let you do that – you'd need a subquery to make the `hits` and `shots` aliases available in other SELECT expressions. I think SQLite lets you do this, but not pgSQL.

Oh well. So a bunch of computed expressions are repeated. Hopefully it's smart enough to optimize.

Reply:
Hmm.. definitely something to look into